Спринт 12

il76 opened the pull request:
Функционально, вроде, работает, но хотелось бы выслушать пожелания по соответствию кодовому стилю Котлина - есть подозрение, что нужно было как-то иначе всё реализовывать...

Commit:
Радиус скругления берём из ресурсного файла.

Commit:
Избавился от dpToPx

Commit:
Класс App для глобальных действий.
Переключение тем перенёс в него.
Сохранение статуса свитчера при рестарте активити

Commit:
Сохраняем статус свитчера в SharedPreferences

Commit:
Заголовок и кнопка очистки истории поиска

Commit:
Обработчик onClick и частично нерабочий код, чтобы не потерять

Commit:
Сохранение истории поиска в SharedPreferences

kekulta commented on app/src/main/java/com/il76/playlistmaker/App.kt:
```diff
@@ -0,0 +1,53 @@
+
+    val trackListHistory = arrayListOf<Track>()
+
+    lateinit var trackSearchHistory: TrackSearchHistory
```
lateinit лучше не использовать лишний раз. Если что-то может инициализироваться в конструкторе, а не в `onCreate`, то оно должно инициализироваться там.

lateinit это костыль. Он ломает логику языка и делает его менее безопасным.
Добавлена эта штука для интеропа с Джавой, например, некоторые функции в фреймворке Spring не работали без этого.

Если где-то есть лейтинит, то это повод очень сильно задуматься.
Почти всегда окажется, что он там либо вообще не нужен, либо можно переписать с использованием делегата `by lazy`.
Если совсем никак, то nullable предпочительнее, т.к. заставляет следить за руками.
В каких-то случаях лейтинит оправдан, например, переменные для View, если не используешь ViewBinding.

Так же сам ViewBinding можно сделать в lateinit, но предпочтительнее использовать конструкцию вроде:
```
private val _binding: ActvityBinding? = null
private val binding: ActvityBinding  get() = requireNotNull(_binding) { "Binding wasn't initiliazed!" }
```

Это нужно, чтобы ошибка была более чёткой.
Так же, например, если использовать binding внутри фрагментов, то переменную нужно ставить null в `onViewDestroyed()` иначе возможна утечка памяти.
При использовании Binding внутри Actvity это не обязательно.

В новом Котлине появилась фича `Backing Fields`, которая сделает эту конутрукцию менее многословной, но она пока что в бете.

kekulta commented on app/src/main/java/com/il76/playlistmaker/App.kt:
```diff
@@ -0,0 +1,53 @@
+
+    lateinit var sharedPrefs: SharedPreferences
+
+    companion object {
+        const val PLAYLIST_MAKER_PREFERENCES = "playlist_maker_preferences"
+        const val DARK_THEME_ENABLED = "dark_theme_enabled"
+        const val TRACKS_SEARCH_HISTORY = "tracks_search_history"
+        lateinit var instance: App
+            private set
+    }
+
```
Согласно [код-стайлу](https://kotlinlang.org/docs/coding-conventions.html#class-layout) Котлина Companion Object должен находится в конце класса.
Хотя я лично знаю команды, где он пишется в начале класса.
В целом, код-стайл это не про жёсткие правила, а про договорённости внутри
команды. Но я бы рекомендовал использовать стандартные практики в новых
проектах, где ещё не устоялись нестадартные. Это будет очень полезно при
онбординге новых разработчиков.

В целом рекомендую внимательно ознакомиться с Coding Conventions Котлина и по
возможности им следовать.

kekulta commented on app/src/main/java/com/il76/playlistmaker/SearchActivity.kt:
```diff
@@ -168,6 +179,50 @@ class SearchActivity : AppCompatActivity() {
+        }
+        toggleSearchHistory(false)
+
+//        Пока вроде он не нужен
+//        spListener = SharedPreferences.OnSharedPreferenceChangeListener { sharedPreferences, key ->
+//            if (key == App.TRACKS_SEARCH_HISTORY) {
+//                val fact = App.instance.sharedPrefs?.getString(App.TRACKS_SEARCH_HISTORY, null)
+//                if (fact != null) {
+//                    adapter.facts.add(0, createFactFromJson(fact))
+//                    adapter.notifyItemInserted(0)
+//                }
+//            }
+//        }
+
+//        App.instance.sharedPrefs.registerOnSharedPreferenceChangeListener(spListener)
```
Закоментированного кода в ПР быть не должно.
Т.к. ты используешь гит, то все предыдущие версии файлов сохранены там. Если ты удалил что-то и хочешь вернуть, то это всегда можно достать оттуда.
Закоментированный код замусоривает кодобаза и делает её менее читабельной. В больших командах так же очень просто потерять ответсвенного, который понимает нужен этот код ещё или уже нет.

Commit:
Кнопка прилеплена к низу списка, а не низу экрана

Commit:
При начале набора текста чистим список треков. В ТЗ поведение не определено, но так нагляднее, как мне кажется.

Commit:
Переделал onclick на нормальный вариант.
Убрал лишние методы из App, а загрузку из SharedPreferences вынес в конструктор вспомогательного класса.

Commit:
Пересортировка истории по клику на элементе в реальном времени
